project/02-Modularity/Handout.md

```diff
@@ -0,0 +1,160 @@
+
+> :information_source: When you design a service that _waits_ on something (for example: an external API), you usually don't want to block the caller, but rather use asynchronous structures. We will introduce them later in the course, so **we don't ask you to use them for now**. **However, remember that -outside of this exercise- calling a web service synchronously from the main thread is in most of the cases a very bad practice.** 
+
+## Implement the interfaces
+
+Now that you wrote the interfaces for your modules, you need to implement them. For each interface, create a class that implements the methods you defined. Here are a few hints and tips on what to do precisely. Feel free to ask on Piazza if you need suggestions!
+
+> :information_source: You may notice while writing the implementation that the interface misses something: a method parameter, a list of thrown exceptions... This is not that big of a deal: sometimes, when writing the actual code, you come up with something you didn't think you'd need. Updating the interface at this time to provide what you need for that is not an issue. **However, please go through the same reasoning every time you update the interface as when you wrote it: is it specific to my implementation? Does it give too much details about the implementation? Do I really need it? ...**. 
+
+> :warning: In particular in today's project, if you realize you need a `Context` (or similar Android system classes) to do something, you should not add it as a method parameter in the interface, as it makes it tightly coupled to the Android operating system. Think about other ways to provide the `Context` to your implementation. 
+
+> :information_source: A `Context` is an interface to global information about an application environment. This is an abstract class whose implementation is provided by the Android system. It allows access to application-specific resources and classes, as well as up-calls for application-level operations such as launching activities, broadcasting and receiving intents, etc. (Source: [Android Documentation](https://developer.android.com/reference/android/content/Context))
+
+### HTTP requests
+
+There are a few options to access the network in Android. We suggest you to use the native `HttpsUrlConnection`. See the [Android Documentation](https://developer.android.com/training/basics/network-ops/connecting#httpsurlconnection) for a good example of how it works.
+
+But first, you need to disable a check. By default, Android disallows doing networking on main thread - as we explained, doing that is very bad. But since we will introduce asynchronicity later, you must add this code somewhere (for example in your main activity). This needs to run before the first call to your service.
```

Comment:
By default, Android disallows doing networking on main thread 
->
By default, Android disallows doing networking on the main thread